Order relationship select

skeary opened the pull request:
Updates the admin UI for the relation type fields so that the values in the select are ordered (ascending) by the label field. This makes it easier to find an option when dealing with larger select lists.

The behavior could be made configurable within `schema.ts` so that ordering could be done on another field or descending rather than ascending order. I wasn't sure of the value of this so I've not made it configurable but explore this if felt useful.

dcousens commented:
Hi @skeary ,

This is nice to have functionality!
Unfortunately,  we don't necessarily want to assume what might be a natural order for this field generally.
We might want to expand this functionality to include configuration in the schema to what the default ordering is,  and use that instead.

If you want to try and do that,  awesome!  Otherwise we'll try and expand on this pull request when we have time :+1:

skeary commented:
Hi @dcousens,
Thanks for the feedback. My thinking was that this was a better default - while it may not solve all use cases it wouldn't be worse for anyone and in general would be what most people wanted? It's not a problem though to add in extra configuration, and it's definitely more flexible, so am happy to have a go at this!

I think the following might be a reasonable configuration approach but just wanted to checked what you thought before I go ahead:
```
    ui?: {
        sort?: 'none' | 'labelField' | { field: string } | undefined;
        sortOrder?: 'asc' | 'desc' | undefined;
    } | undefined;
```

With examples of use in schema.ts:

```
      author: relationship({ ref: 'Author.posts', many: false }), // order on label field, ascending
      author: relationship({ ref: 'Author.posts', many: false, ui: { sort: 'labelField', sortOrder: 'asc' } }), // same
      author: relationship({ ref: 'Author.posts', many: false, ui: { sort: 'none' } }), // no ordering
      author: relationship({ ref: 'Author.posts', many: false, ui: { sort: { field: 'email' }  } }), // order on email, ascending
      author: relationship({ ref: 'Author.posts', many: false, ui: { sort: { field: 'email' }, sortOrder: 'desc'  } }), // order on email, descending
```

What do you think?

skeary commented:
Note this changeset now (I think) includes a working fix for #6112

skeary commented:
Note this changeset now (I think) includes a working fix for https://github.com/keystonejs/keystone/issues/6112. This issue was that the ui.labelField option under the relationship field config wasn't having any affect. I noticed this problem while testing my changes and have now hopefully fixed the issue. Unless I'm missing something, the support for this feature wasn't really there rather than a logic bug or similar. So I've now implemented support for this.

skeary commented:
Hi @dcousens,
I've implemented the approached I outlined in a earlier comment where the configuration now supports:

```
 ui?: {
        orderBy: false | Array<{ labelField:  'asc' | 'desc' } | { field: string, order: 'asc' | 'desc' }> | undefined,
    } | undefined;
```
To me, this seems a reasonably simple way to configure the ordering and should hopefully cover the significant number of use cases. If someone doesn't want ordering they can turn off the ordering which will revert to the current behaviour. By default the ordering is still based on label field (ascending) and I think that is what is useful for most people?

I've updated the documentation to describe the new configuration options. See https://keystone-next-docs-ahungkcd8-keystonejs.vercel.app/docs/apis/fields#relationship for a rendered version.

As part of doing this changeset, I also think I've addressed #6112. 

It would be great if you could review my overall approach and implementation and let me know what you think. I'm happy to tweak as needed.

Cheers,
Simon

willemmulder commented:
Nice work @skeary ; let's hope this can be merged soon!

stuible commented:
Would love the ability to sort relationship select and cards.  Excited to see this feature merged!

dcousens commented:
Replaced by https://github.com/keystonejs/keystone/pull/9648
